pkg/controller/deployment/deployment_controller.go

```diff
@@ -576,6 +579,73 @@ func (dc *DeploymentController) syncRollingUpdateDeployment(deployment extension
 	return dc.syncDeploymentStatus(allRSs, newRS, deployment)
 }
 
+// proportionalScaling should run only on scaling events and not during the normal rollout process.
+func (dc *DeploymentController) proportionalScaling(deployment extensions.Deployment, shouldCreate bool) error {
+	newRS, oldRSs, err := dc.getAllReplicaSets(deployment, shouldCreate)
+	if err != nil {
+		return err
+	}
+	allRSs := controller.FilterActiveReplicaSets(oldRSs)
+	maybeMaxSurge := 0
+	if deployment.Spec.Strategy.Type == extensions.RollingUpdateDeploymentStrategyType {
+		maybeMaxSurge, err = intstrutil.GetValueFromIntOrPercent(&deployment.Spec.Strategy.RollingUpdate.MaxSurge, deployment.Spec.Replicas)
+		if err != nil {
+			return err
+		}
+	}
+
+	if newRS != nil {
+		// No old replica sets with pods; scale new replica set to the full count.
+		if len(controller.FilterActiveReplicaSets(oldRSs)) == 0 {
+			if _, _, err := dc.scaleReplicaSetAndRecordEvent(newRS, deployment.Spec.Replicas, deployment); err != nil {
+				return err
+			}
+			deployment.Status.EnactedReplicas = deployment.Spec.Replicas + maybeMaxSurge
```

Comment:
The point of EnactedReplicas is to detect scaling events, yes, but this adds in maybeMaxSurge before maybeMaxSurge is normalized. I'm concerned this will be confusing to users.

Scenarios:

RecreateUpdate with only one RS with non-zero replicas, not necessarily the new one, in the case that the template doesn't match: If RS replicas != deployment replicas, then it's a scaling event and that RS should be scaled. Is it possible for all RS replicas to be 0 due to pausing in the middle of a recreate?

RollingUpdate with only one RS, not necessarily the new one: The same behavior as Recreate.

RollingUpdate with multiple RSes with non-zero replicas: This is where we need proportional scaling. There are a few subcases where we want to detect a scaling event:
* Sum of RS.Spec.Replicas was Deployment.Spec.Replicas + maybeMaxSurge
* Sum of RS.Spec.Replicas was not Deployment.Spec.Replicas + maybeMaxSurge because newRS.Spec.Replicas already reached Deployment.Spec.Replicas
* Sum of RS.Spec.Replicas was not Deployment.Spec.Replicas + maybeMaxSurge because the controller failed between scaling down the oldRSes and scaling up newRS at the point the Deployment was paused
* Sum of RS.Spec.Replicas was not Deployment.Spec.Replicas + maybeMaxSurge because the controller failed while scaling proportionally after a previous scaling event while paused

If a write to Deployment status fails, all bets are off, unless we can figure out a value to stash in annotations on the RSes, but we should at least be able to handle the first two cases.

What if we stored the fraction of the Deployment and Deployment.Spec.Replicas in annotations on each RS?